### PR TITLE
Return resulting `URI` from commands that save the active editor (fix #178713)

### LINKED ISSUE #178713: Text document `save()` API closes the Untitled document which is saved
Does this issue occur when all extensions are disabled?: Yes

Version: 1.77.0-insider (user setup)
Commit: 7f329fe6c66b0f86ae1574c2911b681ad5a45d63
Date: 2023-03-29T09:56:00.839Z
Electron: 19.1.11
Chromium: 102.0.5005.196
Node.js: 16.14.2
V8: 10.2.154.26-electron.0
OS: Windows_NT x64 10.0.22518
Sandboxed: Yes

Steps to Reproduce:

1. Checkout the `save()` API here: https://code.visualstudio.com/api/references/vscode-api#TextDocument
2. Have an untitled file open, and attempt to save the document using the API
3. `Textdocument.isClosed` property is set to `true`

Also `TextDocument.uri` object is not updated to point to the saved file URI. For Python extension it leads to issues like the following where wrong URI is sent to terminal, also notice that the file disappears:

screen-capture

### PR DESCRIPTION
**Pull request by bpasero**
Fixes https://github.com/microsoft/vscode/issues/178713, https://github.com/microsoft/vscode/issues/124117, https://github.com/microsoft/vscode/issues/22040

@jrieken mainly 👁️ for the new API commands, I feel `vscode.save` and `vscode.saveAs` are better then asking extensions to use our internal commands, especially since that allows us to explicitly have a typed return type. I decided to not just return `vscode.Uri` but `{ uri: vscode.Uri }` because we might later decide to add the view type to the result as well. 

Conceptually saving an editor can result in a different editor opening because it is possible that saving an untitled editor changes the view type (e.g. saving an untitled file as notebook reopens as notebook).

### CODE REVIEW

**Comment by jrieken** [comment]
Not a fan of an having an API command to save the "active editor" because it's really fishy that is and will always change under you. Something more reliable would be a save-command that takes an URI of the resource that you want to save or save-as

**Comment by bpasero** [comment]
@jrieken changed to require a `URI` for both commands and updated the tests. 

Btw this will also address https://github.com/microsoft/vscode/issues/22040 which asks for a way to force a save to disk even when the file is not dirty.

**Comment by bpasero** [comment]
@jrieken as discussed, changed to new proposed API `workspace.save(uri): URI` and `workspace.saveAs(uri): URI` and added tests. This goes along nice with the existing `workspace.saveAll()`.

File: src/vs/workbench/api/common/extHostWorkspace.ts [hunk 1]
```diff
@@ -557,6 +557,18 @@ export class ExtHostWorkspace implements ExtHostWorkspaceShape, IExtHostWorkspac
...
+	async save(uri: URI): Promise<URI | undefined> {
+		const result = await this._proxy.$save(uri);
+
+		return URI.revive(result);
+	}
+
+	async saveAs(uri: URI): Promise<URI | undefined> {
+		const result = await this._proxy.$saveAs(uri);
+
+		return URI.revive(result);
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHostWorkspace.ts:569]: please make sure to enforce both proposals

Comment (jrieken) [reply src/vs/workbench/api/common/extHostWorkspace.ts:569]: 🤦 didn't see that. checking in extHost.api is best

File: src/vs/workbench/api/browser/mainThreadWorkspace.ts [hunk 2]
```diff
@@ -201,8 +203,34 @@ export class MainThreadWorkspace implements MainThreadWorkspaceShape {
...
+	async $save(uriComponents: UriComponents): Promise<UriComponents | undefined> {
+		const uri = URI.revive(uriComponents);
+
+		const editors = [...this._editorService.findEditors(uri, { supportSideBySide: SideBySideEditor.PRIMARY })];
+		const result = await this._editorService.save(editors, { reason: SaveReason.EXPLICIT, force: true /* force save even when non-dirty */ });
+
+		return firstOrDefault(this.saveResultToUris(result));
+	}
+
+	async $saveAs(uriComponents: UriComponents): Promise<UriComponents | undefined> {
```
Comment (jrieken) [inline src/vs/workbench/api/browser/mainThreadWorkspace.ts:215]: nit: only have `$save` but with `$save(uri, saveAs)`

File: src/vs/workbench/api/browser/mainThreadWorkspace.ts [hunk 3]
```diff
@@ -201,8 +203,34 @@ export class MainThreadWorkspace implements MainThreadWorkspaceShape {
...
+		const uri = URI.revive(uriComponents);
+
+		const editors = [...this._editorService.findEditors(uri, { supportSideBySide: SideBySideEditor.PRIMARY })];
+		const result = await this._editorService.save(editors, { reason: SaveReason.EXPLICIT, saveAs: true });
+
+		return firstOrDefault(this.saveResultToUris(result));
+	}
+
+	private saveResultToUris(result: ISaveEditorsResult): URI[] {
```
Comment (jrieken) [inline src/vs/workbench/api/browser/mainThreadWorkspace.ts:224]: `_` ;-)

File: src/vscode-dts/vscode.proposed.saveEditor.d.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ *--------------------------------------------------------------------------------------------*/
+
+// https://github.com/microsoft/vscode/issues/178713
+
+declare module 'vscode' {
+
+	export namespace workspace {
+
+		/**
+		 * Saves the editor identified by the given resource and returns the resulting resource or `undefined`
```
Comment (jrieken) [inline src/vscode-dts/vscode.proposed.saveEditor.d.ts:13]: maybe make it more clear that save only works when a respective editor is open

**Review by jrieken — APPROVED** [review]

**Comment by bpasero** [comment]
PR review feedback will be addressed in https://github.com/microsoft/vscode/pull/180476

**Pull request merged**